### PR TITLE
EventExecutorGroup lightweight progress metrics

### PR DESCRIPTION
**Pull request by franz1981**
Motivation:

EventExecutorGroup doesn't allow to monitor submitted and completed
tasks.

Modifications:

JCtools QueueProgressIndicators exposes lightweight progression metrics:
if an executor is using a proper tasks queue, it will expose such
metrics for free.
MultithreadedEventExecutorGroup can expose global metrics (collected
using children metrics) or fine-grain ones (per-child).

Result:

An external observer can easily monitor executors utilization
without affecting executor performance.

### CODE REVIEW

**Comment by franz1981** [comment]
It is connected to https://github.com/netty/netty/pull/9025: the implementation must be refined and probably I need to add some tests, but I want to receive some feedback about this feature, given that is mostly for-free while adopting JCtools queues :+1:

**Comment by franz1981** [comment]
I've prepared a simple reporting thread on `BurstCostExecutorsBenchmark` to show a possible usage and that's the output:
```
Duration 1001ms - Submitted 24,007,891 evt - Completed 24,007,879 evt
Duration 1001ms - Submitted 23,947,185 evt - Completed 23,947,314 evt
Duration 1001ms - Submitted 23,982,939 evt - Completed 23,982,888 evt
Duration 1000ms - Submitted 23,661,554 evt - Completed 23,661,474 evt
Duration 1001ms - Submitted 23,251,822 evt - Completed 23,251,874 evt
Duration 1001ms - Submitted 23,004,516 evt - Completed 23,004,467 evt
Duration 1001ms - Submitted 23,775,207 evt - Completed 23,775,221 evt
Duration 1001ms - Submitted 24,089,328 evt - Completed 24,089,317 evt
```

**Comment by normanmaurer** [comment]
This is an interesting idea...

@vkostyukov WDYT ?

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by franz1981** [comment]
@normanmaurer I've introduced an API changes that should make the test to fail: this PR is mostly a WIP to understand if such metrics are interesting (given that are mostly for free :P)

**Comment by vkostyukov** [comment]
We use `pendingTasks` to monitor the business of event-loops in. I'm trying to understand what additional insights these methods can bring to us - I'm not really seeing an immediate value at this point.

@franz1981 Any chance you can help me with the example? What are these could be used for where `pendingTasks` can not?

**Comment by franz1981** [comment]
@vkostyukov pendingTasks will give you anothe metric: which tasks are being submitted but not yet executed (not really, but something like  (submitted - completed) and given that submitted tends to happen in bust, if you sample the pending tasks each seconds is likely to be near 0 most of the times.
These metrics instead provides additional insights as "in the period of time what is the throughput of the submissions" ie submitted(t) - submitted(t-1).
Not sure if the explanation make it clear how much is different.
It will helps to know which event loop receive more works and which ones are fast enough to completed them...

File: common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java [hunk 1]
```diff
@@ -29,6 +29,20 @@
  */
 public interface EventExecutorGroup extends ScheduledExecutorService, Iterable<EventExecutor> {
 
+    /**
+     * Returns the number of submitted tasks: it's safe to be read by different threads.<br>
+     * <p>
+     * If {@code this} don't support this metrics it returns a negative value, likely {@code -1}.
+     */
+    long submittedTasks();
```
Comment (njhill) [inline common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java]: I know this is a POC so maybe mainly for illustration, but I don't think we could add new public interface methods like this in 4.1 since it would be a breaking change.

**Comment by njhill** [comment]
@franz1981 I think this is a great POC and wouldn't harm to add (with the inline comment caveat), however as [commented on the related PR](https://github.com/netty/netty/pull/9025#issuecomment-484658536) most useful imo would be to work backwards from the kinds of parameters that may make sense to tune (i.e. those which have most impact on the efficiency/perf) and from that determine which corresponding metrics would be most useful to capture/expose (which might end up being some derived combo based on these counts somehow for example). That could pave the way for possible adaptive config in future.

In other words, how would regular users use these numbers to concretely determine the number of event loops to have in their group(s), etc.

File: common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java (hunk 1 above)
Comment (vkostyukov) [reply common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java]: I had to deal with the same thing in #8428. We can probably just move these under `SingleThreadEventLoop`, which is an abstract class so won't break binary compatibility.

**Comment by franz1981** [comment]
@njhill 

> In other words, how would regular users use these numbers to concretely determine the number of event loops to have in their group(s), etc.

Not easy to answer; probably allowing the user to inject telemetry in order to get:
- task latencies distribution depending on task type (networking/scheduled/common tasks)
- number of wakeups performed into the event loop and outside ie on task offer
- latency distributions of wakeup time (from a select())
- etc etc 

The main idea is to provide callbacks to allow user to do what they want (while avoing Netty prj to contain external unnecessary dependencies) and *at least* provide some basic useful telemetry eg:
- pending, submitted, completed events
- TODO wakeup performed into the event loops
- etc etc

I would let the users ingest the pills of seme heavyweight telemetry (in the form of some callback as `onSleepCompleted(selectedKeys)`)

**Comment by vkostyukov** [comment]
I can imagine this could be useful for measuring the "throughput" of your IO threads. Having just pending tasks wouldn't be enough for that.

I'm +1 for this we can keep it binary compatible.

**Comment by franz1981** [comment]
I need to change this without affecting public API and add tests: will try to do it this week :+1:

**Comment by normanmaurer** [comment]
@franz1981 what is the status of this one ?

**Comment by franz1981** [comment]
@normanmaurer I still need to expose it without changing API and it could be reviewed. No idea which type of tests I can provide for this, need to think about it...
This week I have a huge backlog at work due to the long honeymoon, but will try to get some free cycles to work on this :+1:

**Comment by carryxyh** [comment]
Hi, this pr looks good. We are currently working on a Netty layer monitoring. In our scenario we need three types of data:

1. The pressure of the Reactor thread (it seems that this PR code can reflect this problem)
2. Reporting of resource leaks (Error log is not enough for us)
3. TCP channel back pressure

Often, there are a lot of things we have to doubt when troubleshooting a problem. This data can help us to rule out the anomalies that exist in the Netty layer. I hope that we can eventually have similar capabilities.

**Comment by franz1981** [comment]
@carryxyh thanks! I still need to take care of not changing the public API: I'll try to work on it by the EOD (that's in 9 hours from now). Let me know If you've already tried this branch for your tests and if the API seems appropriate for your use case (that seems the motivation of this PR) :)

**Comment by carryxyh** [comment]
Hi, I haven't tried it yet. But the motivation for this pr is what we need. I am very much looking forward to its completion. :)

**Comment by franz1981** [comment]
@vkostyukov @carryxyh If you wanna give it a try, it should be able to be used without directly modifying a public API (I've used a "dirty" trick I admit, let's see if is fine ;))

@nitsanw This is making use of something specific of JCTools: any objection/comment/proposal is welcome :)

**Comment by nitsanw** [comment]
@franz1981 LGTM, I would suggest making double sure that the interface works as expected for the relevant queues as it's not a commonly used one. I'll add some tests to JCTools just in case.

File: common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java [hunk 2]
```diff
@@ -27,7 +29,7 @@
  * life-cycle and allows shutting them down in a global fashion.
  *
  */
-public interface EventExecutorGroup extends ScheduledExecutorService, Iterable<EventExecutor> {
```
Comment (njhill) [inline common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java]: This is a breaking change, we can't add new methods to `EventExecutorGroup`

File: common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java [hunk 3]
```diff
@@ -15,6 +15,8 @@
  */
 package io.netty.util.concurrent;
 
+import io.netty.util.internal.UnstableApi;
```
Comment (njhill) [inline common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java]: Unused?

File: common/src/main/java/io/netty/util/concurrent/ObservableTaskConsumer.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ * under the License.
+ */
+package io.netty.util.concurrent;
+
+import io.netty.util.internal.UnstableApi;
+
+@UnstableApi
+public interface ObservableTaskConsumer {
+
+    long NOT_OBSERVABLE_TASKS = -1;
```
Comment (njhill) [inline common/src/main/java/io/netty/util/concurrent/ObservableTaskConsumer.java]: Maybe just `UNSUPPORTED`?

File: common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java [hunk 5]
```diff
@@ -1051,6 +1051,16 @@ public void run() {
         });
     }
 
+    @Override
+    public long submittedTasks() {
+        return PlatformDependent.currentProducerIndex(taskQueue);
```
Comment (njhill) [inline common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java]: Do you think it matters that the semantics of producerIndex versus submittedTasks differ in the general case? In particular for the producerIndex `This value will normally indicate number of elements passed into the queue, but may under some circumstances be a derivative of that figure` ... means that some queue impls might cause incorrect values to be returned here (for example double the number of tasks actually submitted).

Comment (franz1981) [reply common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java]: That's something that has to be consider, but for the most used ones should be fine eg https://github.com/JCTools/JCTools/blob/master/jctools-core/src/main/java/org/jctools/queues/BaseMpscLinkedArrayQueue.java#L441

File: common/src/main/java/io/netty/util/concurrent/ObservableTaskConsumer.java (hunk 4 above)
Comment (franz1981) [reply common/src/main/java/io/netty/util/concurrent/ObservableTaskConsumer.java]: I'm very bad with names: ObservableTaskConsumer too... that's not very inspired :)

File: common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java (hunk 2 above)
Comment (franz1981) [reply common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java]: Incredibly it hasn't detected as breaking change by the checks...

Comment (njhill) [reply common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java]: @franz1981 I think you managed to outsmart the checks by adding a new superinterface :)

Comment (franz1981) [reply common/src/main/java/io/netty/util/concurrent/EventExecutorGroup.java]: Argh! Will fix this today

**Comment by franz1981** [comment]
@njhill @normanmaurer I'm now looking if I should add some tests for this, although Nitsan has already added JCTools/JCTools@543ea11: let me know if you prefer better naming or other stuff :)

**Comment by franz1981** [comment]
@njhill @normanmaurer 
AFAIK now it shouldn't break any (broken) compatibility check and we already have some checks on JCTools re `QueueProgressIndicators`: the only thing that's lacking are tests, do you think are needed for this?

**Comment by normanmaurer** [comment]
@franz1981 will review soon.. sorry for the delay.

**Comment by franz1981** [comment]
@normanmaurer np I've taken lot of time as well to get here too :) 
OT: when the PR of Nitsan re JCTools 3 will land I promise to start some experiment to see if there is some benefit using the new xadd qs on Netty ;)

**Comment by franz1981** [comment]
@normanmaurer how it looks? I mean, it's breaking some compatibility now?
Probably need to add some test on this, if it got accepted :)

**Comment by normanmaurer** [comment]
@franz1981 can you rebase it ?

**Comment by fedorovr** [comment]
Oh, seems this PR is stale. @franz1981 Would you mind to update it? Or I will be happy to help with it

File: common/src/main/java/io/netty/util/internal/ObservableTaskConsumerUtil.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        final ObservableTaskConsumer consumer = (ObservableTaskConsumer) obj;
+        return consumer.completedTasks();
+    }
+
+    public static long submittedTasks(Object obj) {
+        if (!(obj instanceof ObservableTaskConsumer)) {
+            return ObservableTaskConsumer.UNSUPPORTED;
+        }
+        final ObservableTaskConsumer consumer = (ObservableTaskConsumer) obj;
+        return consumer.completedTasks();
```
Comment (fedorovr) [inline common/src/main/java/io/netty/util/internal/ObservableTaskConsumerUtil.java:38]: Seems that is copying error, here we need return `consumer.submittedTasks()`?

Comment (fedorovr) [inline common/src/main/java/io/netty/util/internal/ObservableTaskConsumerUtil.java:38]:
```suggestion
        return consumer. submittedTasks();
```

**Comment by franz1981** [comment]
> Oh, seems this PR is stale. @franz1981 Do you have plans to update it? Or I will be happy to help with it

No plans yet and would be happy if you would like to pick it :)
